File: python3/reverse_vowels_of_a_string.py
"""
Space   : O(1)
Time    : O(n)
"""


class Solution:
    def reverseVowels(self, s: str) -> str:
        b, e = 0, (len(s)-1)

        s = list(s)
        while b < e:
            b_get, e_get = False, False

            if s[b].lower() in ['a', 'e', 'i', 'o', 'u']:
                b_get = True

            if s[e].lower() in ['a', 'e', 'i', 'o', 'u']:
                e_get = True

            if b_get and e_get:
                temp = s[b]
                s[b] = s[e]
                s[e] = temp
                b += 1
                e -= 1
            if not b_get:
                b += 1
            if not e_get:
                e -= 1

        return ''.join(s)

File: golang/reverse_vowels_of_a_string.go
package main

/*
Space	: O(n)
Time	: O(n)
*/

func reverseVowels(s string) string {
	a := []byte(s)
	vowels := map[byte]int{
		byte('a'): 1,
		byte('e'): 1,
		byte('i'): 1,
		byte('o'): 1,
		byte('u'): 1,
		byte('A'): 1,
		byte('E'): 1,
		byte('I'): 1,
		byte('O'): 1,
		byte('U'): 1,
	}

	head, tail := 0, len(s)-1
	head_found, tail_found := false, false
	for head < tail {

		if !head_found {
			if _, ok := vowels[s[head]]; ok {
				head_found = true
			} else {
				head++
			}
		}

		if !tail_found {
			if _, ok := vowels[s[tail]]; ok {
				tail_found = true
			} else {
				tail--
			}
		}

		if head_found && tail_found {
			a[head], a[tail] = a[tail], a[head]
			head_found, tail_found = false, false
			head++
			tail--
		}

	}

	return string(a)
}
